tests/TestCase/Middleware/AuthorizationMiddlewareTest.php

```diff
@@ -255,6 +256,38 @@ public function testCustomIdentity()
         $this->assertEquals(1, $result->getAttribute('user')['id']);
     }
 
+    public function testCustomIdentityDecorator()
+    {
+        $identity = $this->createMock(Identity::class);
+
+        $service = $this->createMock(AuthorizationServiceInterface::class);
+        $request = (new ServerRequest)->withAttribute('identity', $identity);
+        $response = new Response();
+        $next = function ($request) {
+            return $request;
+        };
+
+        $middleware = new AuthorizationMiddleware($service, [
+            'identityDecorator' => function ($service, $identity) {
+                $identity->setService($service);
+
+                return $identity;
+            },
+            'requireAuthorizationCheck' => false,
+        ]);
+
+        $identity->expects($this->once())
+            ->method('setService')
+            ->with($service);
```

Comment:
Instead of mocking your stub class you could implement the actual method and rely on fewer mocks.